### PR TITLE
Make non-breaking changes to support arrays of primitives

### PR DESCRIPTION
**Pull request by tgoyne**
A bunch of individually not very interesting changes made to support arrays of primitives. Should have no functional changes other than some error messages being a bit different.

Pulls in https://github.com/realm/realm-object-store/pull/464 (and so depends on that getting merged).

### CODE REVIEW

File: Realm/RLMProperty.mm [hunk 1]
```diff
@@ -354,42 +373,40 @@ - (instancetype)initSwiftPropertyWithName:(NSString *)name
 
     // convert array types to objc variant
     if ([rawType isEqualToString:@"@\"RLMArray\""]) {
-        rawType = [NSString stringWithFormat:@"@\"RLMArray<%@>\"", [propertyValue objectClassName]];
+        RLMArray *value = propertyValue;
+        _type = RLMPropertyTypeArray;
+        _objectClassName = value.objectClassName;
+        if (_objectClassName) {
+            // FIXME: validate
```
Comment (bdash) [inline Realm/RLMProperty.mm]: Is this an existing FIXME that was moved around, or something new that should be addressed?

File: Realm/Tests/DynamicTests.m [hunk 2]
```diff
@@ -161,7 +161,7 @@ - (void)testDynamicTypes {
 
     // verify properties
     RLMRealm *dyrealm = [self realmWithTestPathAndSchema:nil];
-    RLMResults *results = [dyrealm allObjects:AllTypesObject.className];
+    RLMResults<RLMObject *> *results = [dyrealm allObjects:AllTypesObject.className];
```
Comment (bdash) [inline Realm/Tests/DynamicTests.m]: Are these changes required by this PR, or are they being made preemptively in advance of #5149?

File: Realm/Tests/Swift/SwiftSchemaTests.swift [hunk 3]
```diff
@@ -82,13 +82,13 @@ class SwiftRecursingSchemaTestObject : RLMObject {
 }
 
 class InitAppendsToArrayProperty : RLMObject {
-    @objc dynamic var propertyWithIllegalDefaultValue: RLMArray = {
+    @objc dynamic var propertyWithIllegalDefaultValue: RLMArray<SwiftIntObject> = {
```
Comment (bdash) [inline Realm/Tests/Swift/SwiftSchemaTests.swift]: Are these changes required by this PR, or are they being made preemptively in advance of #5149?

File: RealmSwift/Migration.swift [hunk 4]
```diff
@@ -80,7 +80,7 @@ extension Realm {
  instance provides access to the old and new database schemas, the objects in the Realm, and provides functionality for
  modifying the Realm during the migration.
  */
-public final class Migration {
+public struct Migration {
```
Comment (bdash) [inline RealmSwift/Migration.swift]: I assume that changing from class to struct doesn't in any way change how users interact with these types (i.e., it's not a breaking change)?

File: Realm/Tests/DynamicTests.m (hunk 2 above)
Comment (tgoyne) [inline Realm/Tests/DynamicTests.m]: All of the tests build and pass with no changes to the tests other than updating what error message a few of the tests are looking for.

File: RealmSwift/Migration.swift (hunk 4 above)
Comment (tgoyne) [inline RealmSwift/Migration.swift]: I don't see any way for it to break anything, since they're immutable types which the user can't construct directly. `Realm` should maybe also be a struct but that would actually be a breaking change.

Comment (austinzheng) [inline RealmSwift/Migration.swift]: Users can theoretically do stupid things like retroactively conforming `Migration` to a class-only protocol, but I don't think that's something we should worry about.

File: Realm/RLMObjectSchema.mm [hunk 5]
```diff
@@ -114,7 +114,8 @@ + (instancetype)schemaForObjectClass:(Class)objectClass {
     Class superClass = class_getSuperclass(cls);
     NSArray *allProperties = @[];
     while (superClass && superClass != RLMObjectBase.class) {
-        allProperties = [[RLMObjectSchema propertiesForClass:cls isSwift:isSwift] arrayByAddingObjectsFromArray:allProperties];
+        allProperties = [[RLMObjectSchema propertiesForClass:cls isSwift:isSwift]
+                         arrayByAddingObjectsFromArray:allProperties];
```
Comment (jpsim) [inline Realm/RLMObjectSchema.mm]: indentation

File: Realm/RLMObjectSchema.mm [hunk 6]
```diff
@@ -217,87 +218,81 @@ + (NSArray *)propertiesForClass:(Class)objectClass isSwift:(bool)isSwiftClass {
...
+            if (!type) {
+                return;
+            }
+            if (property) {
+                property.type = RLMPropertyType(type.intValue);
+                return;
+            }
+
+            // Check to see if this optional property is an underlying storage property for a Swift lazy var.
+            // Managed lazy vars are't allowed.
```
Comment (jpsim) [inline Realm/RLMObjectSchema.mm]: "aren't"
